resolver.go

```diff
@@ -34,6 +35,47 @@ type Resolver struct {
 	cache   *cache
 	expire  bool
 	timeout time.Duration
+	dialer  *net.Dialer
+}
+
+// ResolverOption configures a single resolver setting
+type ResolverOption func(r *Resolver)
```

Comment:
This should be an interface, or a `struct` with no exported fields. If an option can fail, the apply method should return an error, which should be propagated to `NewResolver`.

Reply:
I looked through a few Function Options examples and they all either use the function signature directly or declare a type mapping to a function.
Can you give me a quick example on what it should look like as an interface/struct type? Or a link to what you had in mind?

Reply:
Roughly like this:

```go
type ResolverOption resolverOption

type resolverOption func(*Resolver) error
```

Or using an interface:

```go
type ResolverOption resolverOption

type resolverOption interface {
	apply(*Resolver) error
}

type resolverOptionFunc func(*Resolver) error

func (o resolverOptionFunc) apply(*Resolver) error {
	return nil
}
```

Reply:
Ah great, I already made the func type private. So adding the error return and adding a public alias, will do!

Reply:
found an few minutes of :bullettrain_side: time to do this, I think all comments are now addressed, would you give it another check?